File: ios/BlobChannelModule.swift
import Foundation
import React

/// The fast transfer path, as this iPhone implements it.
///
/// JavaScript decides whether to use it and mints the key material; this module moves the bytes. Both
/// halves stream with the cipher inline, so a model larger than the phone's memory transfers without
/// ever being held in it, and the thread that draws the screen never sees a byte.
///
/// Progress arrives as an event rather than in the promise, because the point of progress is to be
/// visible while the work is still running.
@objc(BlobChannelModule)
final class BlobChannelModule: RCTEventEmitter {
  private static let progressEvent = "SyncBlobProgress"
  private static let outcomeEvent = "SyncBlobOutcome"

  private lazy var server = BlobChannelServer(
    onProgress: { [weak self] requestId, bytes in
      self?.report(requestId: requestId, bytes: bytes)
    },
    onOutcome: { [weak self] requestId, landed in
      guard let self, self.bridge != nil else { return }
      self.sendEvent(
        withName: Self.outcomeEvent, body: ["requestId": requestId, "landed": landed])
    })
  private let work = DispatchQueue(label: "ai.offgridmobile.blob-channel-module", attributes: .concurrent)

  override static func requiresMainQueueSetup() -> Bool { false }

  override func supportedEvents() -> [String] { [Self.progressEvent, Self.outcomeEvent] }

  /// The address every native sync listener on this phone can accept connections on.
  @objc(lanAddress:withRejecter:)
  func lanAddress(
    resolve: @escaping RCTPromiseResolveBlock,
    reject _: @escaping RCTPromiseRejectBlock
  ) {
    work.async { resolve(BlobChannelSupport.lanAddress()) }
  }

  /// All current IPv4 interfaces; the shared QR projector decides which routes are safe.
  @objc(interfaceCandidates:withRejecter:)
  func interfaceCandidates(
    resolve: @escaping RCTPromiseResolveBlock,
    reject _: @escaping RCTPromiseRejectBlock
  ) {
    work.async {
      resolve(BlobChannelSupport.interfaceCandidates().map { candidate in
        ["host": candidate.host, "interfaceName": candidate.interfaceName]
      })
    }
  }

  /// Offer an endpoint for one transfer, and answer the url a peer should stream to.
  ///
  /// Resolves with nothing when this device has no address on a shared network: there is no endpoint
  /// to offer, and the caller falls back to the path that always works.
  @objc(serve:resolve:withRejecter:)
  func serve(
    _ options: NSDictionary,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    work.async {
      guard let requestId = options["requestId"] as? String,
        let token = options["token"] as? String,
        let destination = options["destinationPath"] as? String,
        let key = Self.decode(options["keyBase64"]),
        let nonce = Self.decode(options["nonceBase64"])
      else { return reject("blob_channel_failed", "the transfer is missing its material", nil) }
      guard let address = BlobChannelSupport.lanAddress() else { return resolve(nil) }
      do {
        let port = try self.server.ensureListening()
        self.server.offer(
          requestId: requestId,
          transfer: .init(
            token: token,
            destinationPath: destination,
            fileSize: (options["fileSize"] as? NSNumber)?.intValue ?? 0,
            key: key,
            nonce: nonce,
            frameBytes: (options["frameBytes"] as? NSNumber)?.intValue ?? 0,
            offset: (options["offset"] as? NSNumber)?.intValue ?? 0,
            expiresAt: Date().addingTimeInterval(
              ((options["ttlMs"] as? NSNumber)?.doubleValue ?? 0) / 1000)))
        let encoded =
          requestId.addingPercentEncoding(withAllowedCharacters: .alphanumerics) ?? requestId
        resolve(["url": "http://\(address):\(port)/blob/\(encoded)"])
      } catch {
        reject("blob_channel_failed", error.localizedDescription, error)
      }
    }
  }

  /// Stop serving an endpoint, whether its transfer completed or not.
  @objc(release:)
  func release(_ requestId: String) {
    server.release(requestId: requestId)
  }

  /// Stop sending a payload that is still going out.
  @objc(abort:)
  func abort(_ requestId: String) {
    BlobChannelUploader.abort(requestId)
  }

  /// Send a local file through the endpoint a peer offered, sealing it on the way out.
  @objc(stream:resolve:withRejecter:)
  func stream(
    _ options: NSDictionary,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    work.async {
      guard let requestId = options["requestId"] as? String,
        let source = options["sourcePath"] as? String,
        let text = options["url"] as? String, let url = URL(string: text),
        let token = options["token"] as? String,
        let key = Self.decode(options["keyBase64"]),
        let nonce = Self.decode(options["nonceBase64"])
      else { return reject("blob_channel_failed", "the transfer is missing its material", nil) }
      do {
        let sent = try BlobChannelUploader.upload(
          .init(
            requestId: requestId, sourcePath: source, url: url, token: token, key: key,
            nonce: nonce, frameBytes: (options["frameBytes"] as? NSNumber)?.intValue ?? 0,
            offset: (options["offset"] as? NSNumber)?.intValue ?? 0)
        ) { [weak self] bytes in
          self?.report(requestId: requestId, bytes: bytes)
        }
        resolve(["bytes": sent])
      } catch {
        reject("blob_channel_failed", error.localizedDescription, error)
      }
    }
  }

  private func report(requestId: String, bytes: Int) {
    guard bridge != nil else { return }
    sendEvent(withName: Self.progressEvent, body: ["requestId": requestId, "bytes": bytes])
  }

  private static func decode(_ value: Any?) -> Data? {
    guard let text = value as? String else { return nil }
    return Data(base64Encoded: text)
  }
}

File: android/app/src/main/java/ai/offgridmobile/sync/BlobChannelModule.kt
package ai.offgridmobile.sync

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import java.util.concurrent.Executors

/**
 * The fast transfer path, as this phone implements it.
 *
 * JavaScript decides WHETHER to use it and mints the key material; this module does the moving. One
 * native call per direction, each streaming disk to socket with the cipher inline - so a model larger
 * than the phone's RAM transfers without ever being held in it, and the thread that draws the screen
 * never sees a byte.
 *
 * Progress comes back as an event rather than a promise, because the point of it is to arrive while
 * the work is still going.
 */
class BlobChannelModule(
    private val context: ReactApplicationContext,
) : ReactContextBaseJavaModule(context) {
    override fun getName(): String = NAME

    private val work = Executors.newCachedThreadPool()
    private val server by lazy { BlobServer(::emitProgress, ::emitOutcome) }

    /** The address every native sync listener on this phone can accept connections on. */
    @ReactMethod
    fun lanAddress(promise: Promise) {
        work.execute { promise.resolve(BlobCrypto.lanAddress()) }
    }

    /** All current IPv4 interfaces; the shared QR projector decides which routes are safe. */
    @ReactMethod
    fun interfaceCandidates(promise: Promise) {
        work.execute {
            val result = Arguments.createArray()
            BlobCrypto.interfaceCandidates().forEach { candidate ->
                result.pushMap(
                    Arguments.createMap().apply {
                        putString("host", candidate.host)
                        putString("interfaceName", candidate.interfaceName)
                    },
                )
            }
            promise.resolve(result)
        }
    }

    /**
     * Offer an endpoint for one transfer, and answer the url a peer should stream to.
     *
     * Returns nothing when this device has no address on a shared network - there is no endpoint to
     * offer, and the caller falls back to the path that always works.
     */
    @ReactMethod
    fun serve(options: ReadableMap, promise: Promise) {
        work.execute {
            try {
                val requestId = options.requireText("requestId")
                val address = BlobCrypto.lanAddress()
                if (address == null) {
                    promise.resolve(null)
                    return@execute
                }
                val port = server.ensureListening()
                server.offer(
                    requestId,
                    BlobServer.Pending(
                        token = options.requireText("token"),
                        destinationPath = options.requireText("destinationPath"),
                        fileSize = options.getDouble("fileSize").toLong(),
                        keyBase64 = options.requireText("keyBase64"),
                        nonceBase64 = options.requireText("nonceBase64"),
                        frameBytes = options.getDouble("frameBytes").toInt(),
                        offset = options.getDouble("offset").toLong(),
                        expiresAt = System.currentTimeMillis() + options.getDouble("ttlMs").toLong(),
                    ),
                )
                promise.resolve(
                    Arguments.createMap().apply {
                        putString(
                            "url",
                            "http://$address:$port/blob/${java.net.URLEncoder.encode(requestId, "UTF-8")}",
                        )
                    },
                )
            } catch (error: Exception) {
                promise.reject(BLOB_FAILED, error)
            }
        }
    }

    /** Stop serving an endpoint, whether its transfer completed or not. */
    @ReactMethod
    fun release(requestId: String) {
        server.release(requestId)
    }

    /** Stop sending a payload that is still going out. */
    @ReactMethod
    fun abort(requestId: String) {
        BlobUploader.abort(requestId)
    }

    /** Send a local file through the endpoint a peer offered, sealing it on the way out. */
    @ReactMethod
    fun stream(options: ReadableMap, promise: Promise) {
        work.execute {
            try {
                val requestId = options.requireText("requestId")
                val sent = BlobUploader.upload(
                    BlobUploader.Request(
                        requestId = requestId,
                        offset = options.getDouble("offset").toLong(),
                        sourcePath = options.requireText("sourcePath"),
                        url = options.requireText("url"),
                        token = options.requireText("token"),
                        keyBase64 = options.requireText("keyBase64"),
                        nonceBase64 = options.requireText("nonceBase64"),
                        frameBytes = options.getDouble("frameBytes").toInt(),
                    ),
                ) { bytes -> emitProgress(requestId, bytes) }
                promise.resolve(
                    Arguments.createMap().apply { putDouble("bytes", sent.toDouble()) },
                )
            } catch (error: Exception) {
                promise.reject(BLOB_FAILED, error)
            }
        }
    }

    // Required by the event emitter contract; the listeners live entirely on the JavaScript side.
    @ReactMethod fun addListener(eventName: String) = Unit

    @ReactMethod fun removeListeners(count: Int) = Unit

    private fun emitProgress(requestId: String, bytes: Long) {
        emit(
            PROGRESS_EVENT,
            Arguments.createMap().apply {
                putString("requestId", requestId)
                putDouble("bytes", bytes.toDouble())
            },
        )
    }

    private fun emitOutcome(requestId: String, landed: Boolean) {
        emit(
            OUTCOME_EVENT,
            Arguments.createMap().apply {
                putString("requestId", requestId)
                putBoolean("landed", landed)
            },
        )
    }

    private fun emit(event: String, payload: WritableMap) {
        if (!context.hasActiveReactInstance()) return
        context
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(event, payload)
    }

    private fun ReadableMap.requireText(key: String): String =
        getString(key) ?: throw IllegalArgumentException("$key is required")

    private companion object {
        const val NAME = "SyncBlobChannelModule"
        const val PROGRESS_EVENT = "SyncBlobProgress"
        const val OUTCOME_EVENT = "SyncBlobOutcome"
        const val BLOB_FAILED = "blob_channel_failed"
    }
}
